org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -2005,6 +2010,25 @@ public ReferenceBinding inferRecordPatternParameterization(
 		return getRecordPatternTypeFromUpwardsProjection(typeBinding, alphas, solution);
 	}
 
+	public void addConstraintsFromNestedPatterns(RecordPattern recordPattern) {
+		if (recordPattern.patterns != null) {
+			Pattern[] patterns = recordPattern.patterns;
+			RecordComponentBinding[] components = ((ReferenceBinding) recordPattern.resolvedType.original()).components();
+			List<ConstraintFormula> extendedConstraints = new ArrayList<>();
+			if (this.initialConstraints != null)
+				extendedConstraints.addAll(Arrays.asList(this.initialConstraints));
+			for (int i = 0; i < patterns.length; i++) {
+				Pattern pattern = patterns[i];
+				if (pattern instanceof TypePattern && pattern.resolvedType != null && pattern.resolvedType.isValidBinding()) {
+					TypeBinding compTypeSubst = substitute(components[i].type);
+					if (!compTypeSubst.isProperType(true))
```

Comment:
this check could be wrong, perhaps the constraint needs to be created always, to help detecting incompatible situations.